### ISSUE TITLE
Error while trying to restore from checkpoint after changing nativelstm2 to basiclstm

### ISSUE THREAD

**Issue by manish-kumar-garg** [rwth-i6/returnn#274]
I trained a model using `2018-asr-attention` `full-setup-attention` which uses `nativelstm2`.
However, later realized that I want to use `basiclstm` at the time of inference, so changed the config with `basiclstm` instead of `nativelstm` and built a meta file using below command.
`python returnn/tools/compile_tf_graph.py returnn.basiclstm.config --eval 1 --search 1 --verbosity 5 --output_file out.meta`

Now, I am trying to use this meta file to restore the weights.

Getting key not found error because the names are not same.
```
Traceback (most recent call last):
  File "freeze_graph.py", line 21, in <module>
    saver.restore(sess, input_checkpoint)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1292, in restore
    err, "a Variable name or other graph key that is missing")
tensorflow.python.framework.errors_impl.NotFoundError: Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is mi
ssing from the checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key lstm0_bw/rec/rnn/basic_lstm_cell/bias not found in checkpoint
         [[node save/RestoreV2 (defined at freeze_graph.py:16) ]]

Caused by op 'save/RestoreV2', defined at:
  File "freeze_graph.py", line 16, in <module>
    saver = tf.train.import_meta_graph('out.1.meta',clear_devices=clear_devices)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1435, in import_meta_graph
    meta_graph_or_file, clear_devices, import_scope, **kwargs)[0]
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/training/saver.py", line 1457, in _import_meta_graph_with_return_elements
    **kwargs))
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/meta_graph.py", line 806, in import_scoped_meta_graph_with_return_elements
    return_elements=return_elements)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/util/deprecation.py", line 507, in new_func
    return func(*args, **kwargs)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/importer.py", line 442, in import_graph_def
    _ProcessNewOps(graph)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/importer.py", line 235, in _ProcessNewOps
    for new_op in graph._add_new_tf_operations(compute_devices=False):  # pylint: disable=protected-access
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3433, in _add_new_tf_operations
    for c_op in c_api_util.new_tf_operations(self)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3433, in <listcomp>
    for c_op in c_api_util.new_tf_operations(self)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 3325, in _create_op_from_tf_operation
    ret = Operation(c_op, self)
  File "/home/ubuntu/tf1.13/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1801, in __init__
    self._traceback = tf_stack.extract_stack()

NotFoundError (see above for traceback): Restoring from checkpoint failed. This is most likely due to a Variable name or other graph key that is missing from the
 checkpoint. Please ensure that you have not altered the graph expected based on the checkpoint. Original error:

Key lstm0_bw/rec/rnn/basic_lstm_cell/bias not found in checkpoint
         [[node save/RestoreV2 (defined at freeze_graph.py:16) ]]

```
How can I make sure that the names of the layers is same?

**Comment by albertz** [comment]
Not only the names are different, the parameters themselves are also stored in a different way.

Returnn however can automatically do the conversion for you (param names and also converting different formats). To do that, you need to start Returnn with your new config `returnn.basiclstm.config`, and make it load the existing checkpoint (trained using NativeLstm2, which it will automatically convert), and then save the checkpoint again. I think sth like `rnn.py returnn.basiclstm.config --task initialize_model` would exactly do that. Then just use that new/converted checkpoint.

**Comment by manish-kumar-garg** [comment]
Getting below error while using `basiclstm` 

```
Using gpu device 0: Tesla V100-SXM2-16GB
<LibriSpeechCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 183.267376, new is 63.708029, requested max is 75.000000. Old num seqs is 6575, new num seqs is 822.
<LibriSpeechCorpus 'train' epoch=1>, epoch 1. Old num seqs 14063, new num seqs 822.
<LibriSpeechCorpus 'train' epoch=1>, epoch 1. Old mean seq len (transcription) is 183.267376, new is 63.708029, requested max is 75.000000. Old num seqs is 6575, new num seqs is 822.
<LibriSpeechCorpus 'train' epoch=1>, epoch 1. Old num seqs 14063, new num seqs 822.
Train data:
  input: 40 x 1
  output: {'classes': [10025, 1], 'raw': {'dtype': 'string', 'shape': ()}, 'data': [40, 2]}
  LibriSpeechCorpus, sequences: 822, frames: unknown
Dev data:
  LibriSpeechCorpus, sequences: 3000, frames: unknown
Unhandled exception <class 'AssertionError'> in thread <_MainThread(MainThread, started 140175512801408)>, proc 7858.

Thread current, main, <_MainThread(MainThread, started 140175512801408)>:
(Excluded thread.)

That were all threads.
EXCEPTION
Traceback (most recent call last):
  File "returnn/rnn.py", line 654, in <module>
    line: main(sys.argv)
    locals:
      main = <local> <function main at 0x7f7d1de1d730>
      sys = <local> <module 'sys' (built-in)>
      sys.argv = <local> ['returnn/rnn.py', 'returnn_basiclstm.config', '--task', 'initialize_model'], _[0]: {len = 14}
  File "returnn/rnn.py", line 642, in main
    line: execute_main_task()
    locals:
      execute_main_task = <global> <function execute_main_task at 0x7f7d1de1d620>
  File "returnn/rnn.py", line 569, in execute_main_task
    line: engine.init_train_from_config(config, train_data, dev_data, eval_data)
    locals:
      engine = <global> <TFEngine.Engine object at 0x7f7a5c4a94a8>
      engine.init_train_from_config = <global> <bound method Engine.init_train_from_config of <TFEngine.Engine object at 0x7f7a5c4a94a8>>
      config = <global> <Config.Config object at 0x7f7d264b3978>
      train_data = <global> <LibriSpeechCorpus 'train' epoch=1>
      dev_data = <global> <LibriSpeechCorpus 'dev' epoch=1>
      eval_data = <global> None
  File "/home/ubuntu/manish/returnn-experiments/2018-asr-attention/librispeech/full-setup-attention/returnn/TFEngine.py", line 864, in init_train_from_config
    line: self.start_epoch, self.start_batch = self.get_train_start_epoch_batch(config)
    locals:
      self = <local> <TFEngine.Engine object at 0x7f7a5c4a94a8>
      self.start_epoch = <local> None
      self.start_batch = <local> !AttributeError: 'Engine' object has no attribute 'start_batch'
      self.get_train_start_epoch_batch = <local> <bound method EngineBase.get_train_start_epoch_batch of <class 'TFEngine.Engine'>>
      config = <local> <Config.Config object at 0x7f7d264b3978>
  File "/home/ubuntu/manish/returnn-experiments/2018-asr-attention/librispeech/full-setup-attention/returnn/EngineBase.py", line 154, in get_train_start_epoch_batch
    line: last_epoch, _ = cls.get_epoch_model(config)
    locals:
      last_epoch = <not found>
      _ = <not found>
      cls = <local> <class 'TFEngine.Engine'>
      cls.get_epoch_model = <local> <bound method EngineBase.get_epoch_model of <class 'TFEngine.Engine'>>
      config = <local> <Config.Config object at 0x7f7d264b3978>
  File "/home/ubuntu/manish/returnn-experiments/2018-asr-attention/librispeech/full-setup-attention/returnn/EngineBase.py", line 91, in get_epoch_model
    line: assert load_epoch in existing_models
    locals:
      load_epoch = <local> 80
      existing_models = <local> {}
AssertionError

```

**Comment by manish-kumar-garg** [comment]
This occurs even when starting training from scratch with `basiclstm`.

**Comment by albertz** [comment]
I guess you configured `model` wrongly in the config. It should point to the previous model.
But better: Use `load` instead to explicitly point it to the model you want to load, and then configure `model` to be the new (output) model filename.

**Comment by manish-kumar-garg** [comment]
Ohh.. Yes. My mistake. config was having `load_epoch`. It worked fine now. Thanks!

**Comment by manish-kumar-garg** [comment]
I think, something is going off here. 
Changes I have made are these to config.

1. Changed `nativelstm2` to `basiclstm`
2. Changed `LSTMBlock` to `BasicLSTM`
3. Added `load = "data/exp-returnn/model.180"` to config
4. Added `model = "data/exp-returnn_basiclstm/model"` to config

When running, getting `Key NotFoundError`
Full stdout [here](https://gist.github.com/manish-kumar-garg/61b1863b1f55d7062e0dfac0f301f0de)

Is there anything I should change?

**Comment by albertz** [comment]
I see. I think the automatic conversion got broken in some newer TF version because they (again) changed the variable names. It should work with some older TF version (not exactly sure which though... 1.8 definitely should work, probably also newer, e.g. maybe try 1.12).
I will also try to fix that on the Returnn side for newer TF versions (but probably not until next week or so).

**Comment by manish-kumar-garg** [comment]
Tried with `standardlstm` also. Same error.
I will try with older tf version.

**Comment by mattiadg** [comment]
I get the same error when converting from `nativelstm2` to `standardlstm`.
This is the failing mapping
```
Variables to restore which are not in checkpoint:
lstm0_bw/rec/rnn/lstm_cell/bias
lstm0_bw/rec/rnn/lstm_cell/kernel
lstm0_fw/rec/rnn/lstm_cell/bias
lstm0_fw/rec/rnn/lstm_cell/kernel
lstm1_bw/rec/rnn/lstm_cell/bias
lstm1_bw/rec/rnn/lstm_cell/kernel
lstm1_fw/rec/rnn/lstm_cell/bias
lstm1_fw/rec/rnn/lstm_cell/kernel
lstm2_bw/rec/rnn/lstm_cell/bias
lstm2_bw/rec/rnn/lstm_cell/kernel
lstm2_fw/rec/rnn/lstm_cell/bias
lstm2_fw/rec/rnn/lstm_cell/kernel
lstm3_bw/rec/rnn/lstm_cell/bias
lstm3_bw/rec/rnn/lstm_cell/kernel
lstm3_fw/rec/rnn/lstm_cell/bias
lstm3_fw/rec/rnn/lstm_cell/kernel

Variables in checkpoint which are not needed for restore:
lstm0_bw/rec/W
lstm0_bw/rec/W_re
lstm0_bw/rec/b
lstm0_fw/rec/W
lstm0_fw/rec/W_re
lstm0_fw/rec/b
lstm1_bw/rec/W
lstm1_bw/rec/W_re
lstm1_bw/rec/b
lstm1_fw/rec/W
lstm1_fw/rec/W_re
lstm1_fw/rec/b
lstm2_bw/rec/W
lstm2_bw/rec/W_re
lstm2_bw/rec/b
lstm2_fw/rec/W
lstm2_fw/rec/W_re
lstm2_fw/rec/b
lstm3_bw/rec/W
lstm3_bw/rec/W_re
lstm3_bw/rec/b
lstm3_fw/rec/W
lstm3_fw/rec/W_re
lstm3_fw/rec/b
output/rec/output_prob/W
output/rec/output_prob/b
output/rec/readout_in/W
output/rec/readout_in/b

Probably we can restore these:
(None)
```

**Comment by albertz** [comment]
I added a fix (and test case) which should fix the conversion nativelstm2->standardlstm and also nativelstm2->basiclstm. Can you try?

Reopen if you still have this problem, and please add further information then.
